readthedocs/proxito/views/serve.py

```diff
@@ -195,35 +195,40 @@ def get(self, request, proxito_path, template_name='404.html'):
         storage = get_storage_class(settings.RTD_BUILD_MEDIA_STORAGE)()
 
         # First, check for dirhtml with slash
-        for tryfile in ('index.html', 'README.html'):
-            storage_filename_path = os.path.join(
-                storage_root_path, filename, tryfile
-            )
-            log.debug(
-                'Trying index filename: project=%s version=%s, file=%s',
-                final_project.slug,
-                version_slug,
-                storage_filename_path,
-            )
-            if storage.exists(storage_filename_path):
-                log.info(
-                    'Redirecting to index file: project=%s version=%s, storage_path=%s',
+        doc_type = (
+            Version.objects.filter(project=final_project, slug=version_slug)
+            .values_list('documentation_type', flat=True)
+            .first()
+        )
+        is_htmldir_type = doc_type in (SPHINX_HTMLDIR, MKDOCS)
+        if is_htmldir_type:
```

Comment:
I don't think this check is correct. We have users who want to serve README.html as their index, regardless of their doctype, I don't think we can skip that check. 

I'd like to remove the README check all together though, but I don't think this is the way to do it.

Reply:
Actually, I don't think we can skip the `index.html` check here either. It doesn't depend on htmldir either:

For example we want `https://docs.readthedocs.io/en/latest` to redirect to `https://docs.readthedocs.io/en/latest/`, and it doesn't matter what doctype it is.

Reply:
I thought we did that here

https://github.com/readthedocs/readthedocs.org/blob/33049ccfc70c335d477348788319879721bec520/readthedocs/proxito/views/serve.py#L125-L126

But looks like we send them to the 404 handler

https://github.com/readthedocs/readthedocs.org/blob/33049ccfc70c335d477348788319879721bec520/readthedocs/proxito/urls.py#L122-L133

Is there a reason why we send them to the 404 view and not just redirect them with `/`? (same for subprojects, translations, single page, etc)

But still I can see cases where users want to serve an index without caring about the doctype. That could be, users using a `html`  doctype and having a directory with an index...

I guess we could at least assume `404.html` for `html` types and `404/index.html` for `htmldir` types when serving a custom 404 page? that will give us one check less

Reply:
We have to do a query to cloud storage to know if the file exists before we redirect to `/`, and we don't want to do any queries in the normal proxito handler to make it really fast. The point of the 404 handler is to be slower and handle the queries. 

> I guess we could at least assume 404.html for html types and 404/index.html for htmldir types when serving a custom 404 page? that will give us one check less

Yea, that seems easy enough. Though I could see users hacking in a top-level 404.html` even with a `dirhtml` builder. We should probably keep stats, but I assume we almost never serve `404/index.html`, and could consider removing it.

Reply:
https://www.mkdocs.org/user-guide/deploying-your-docs/#404-pages

Looks like mkdocs always outputs a `404.html` file, not a `404/inde.html` file. And github pages expects a `404.html` https://help.github.com/en/github/working-with-github-pages/creating-a-custom-404-page-for-your-github-pages-site

Sphinx with `htmldir` is the only one that outputs a `404/index.html` file.